### PR TITLE
add openssl/3.0.0

### PR DESCRIPTION
**Pull request by Croydon**
Specify library name and version:  **openssl/3.0.0**

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/wiki#how-to-submit-a-pull-request) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

Continuation of https://github.com/conan-io/conan-center-index/pull/2054
_____________________________

The intention of this is **NOT**:
  * to merge any pre-release versions - all pre-release versions will get removed later
  * to merge any of this before the final 3.0.0 version gets released, which means this will stay a draft for several more months

The intention of this is:
  * to get early feedback
  * to test early on and notice bugs early on
  * to be able to release a recipe for a final 3.0.0 version shortly after its release
  * to get contributions (help wanted!) 😄 

Important resources:
  * https://wiki.openssl.org/index.php/OpenSSL_3.0
  * https://github.com/openssl/openssl/blob/master/NEWS.md#openssl-30
  * https://github.com/openssl/openssl/blob/master/INSTALL.md

Upstream bugs:
  * [x] Builds fails when path contains `@` https://github.com/openssl/openssl/issues/12078

Open questions / TODO (certainly incomplete):
  * [ ] since #1586 we patched OpenSSL for some Apple operating systems which don't have `fork()`. OpenSSL merged in the meantime a "partial fix" (https://github.com/openssl/openssl/commit/ca811248d838058c13236a6c3b688e0ac98c02c8), so even when we set one of the flags (`HAVE_FORK` or `NO_FORK`, which one would it be anyway? Why two of them?) it does not seem to be fixed upstream. But we should probably wait a few pre-releases longer, before we might patch from our side again. See also https://github.com/openssl/openssl/issues/7607
  * [ ] in `package()` when `nmake` is used `build_type` is `Debug` we give the lib files a `d` prefix. Why? Is this really wanted and still needed for >= 3.0.0?
  * [ ] we seem to have several problems with `find_package(OpenSSL)`/`FindOpenSSL.cmake`, see my comments here: https://github.com/conan-io/conan-center-index/pull/2054/commits/a2bb89ba04d90df0d01c3e34daf21485a5cae5d1
  * [x] the `test_package` will probably fail in its current form with certain different option values
  * [x] Usage of environment variable `CROSS_COMPILE` https://github.com/conan-io/conan-center-index/pull/2054#discussion_r523453297

### CODE REVIEW

File: recipes/openssl/3.x.x/conanfile.py [hunk 1]
```diff
@@ -0,0 +1,639 @@
...
+                    "mips": "o32",
+                    "mips64": "64",
+                    "x86": "android",
+                    "x86_64": "elf"}.get(the_arch, None)
+        return None
+
+    @property
+    def _asm_target(self):
+        the_os = str(self.settings.os)
+        if the_os in ["Android", "iOS", "watchOS", "tvOS"]:
```
Comment (madebr) [inline recipes/openssl/3.x.x/conanfile.py:147]: The strinigication is not needed, I think.
Bonus is extra conan checks.
```suggestion
        if self.settings.os in ("Android", "iOS", "watchOS", "tvOS"):
```

File: recipes/openssl/3.x.x/conanfile.py [hunk 2]
```diff
@@ -0,0 +1,639 @@
+import os
+import fnmatch
+import platform
+from conans.errors import ConanInvalidConfiguration, ConanException
+from conans import ConanFile, AutoToolsBuildEnvironment, tools
+from conans.tools import Version
+
+
+class OpenSSLConan(ConanFile):
```
Comment (madebr) [inline recipes/openssl/3.x.x/conanfile.py:9]:
```suggestion
from conans import ConanFile, AutoToolsBuildEnvironment, tools
from conans.errors import ConanInvalidConfiguration, ConanException
import os
import fnmatch
import platform


required_conan_version = ">=1.33.0"


class OpenSSLConan(ConanFile):
```

File: recipes/openssl/3.x.x/test_package/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,18 @@
+cmake_minimum_required(VERSION 3.1)
+project(test_package)
+
+include(${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)
+conan_basic_setup()
```
Comment (madebr) [inline recipes/openssl/3.x.x/test_package/CMakeLists.txt:5]:
```suggestion
conan_basic_setup(TARGETS)
```

File: recipes/openssl/config.yml [hunk 4]
```diff
@@ -23,9 +23,5 @@ versions:
     folder: "1.x.x"
   1.1.1i:
     folder: "1.x.x"
-  1.1.1j:
-    folder: "1.x.x"
-  1.1.1k:
-    folder: "1.x.x"
-  1.1.1l:
-    folder: "1.x.x"
```
Comment (madebr) [inline recipes/openssl/config.yml:31]: Accidental removal here.

File: recipes/openssl/3.x.x/test_package/conanfile.py [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+        if self.settings.os == "Android":
+            cmake.definitions["CONAN_LIBCXX"] = ""
+
+        cmake.definitions["OPENSSL_WITH_ZLIB"] = self.options["openssl"].zlib
+        cmake.definitions["OPENSSL_ROOT_DIR"] = self.deps_cpp_info["openssl"].rootpath
+        cmake.definitions["OPENSSL_USE_STATIC_LIBS"] = not self.options["openssl"].shared
+        if self.settings.compiler == 'Visual Studio':
+            cmake.definitions["OPENSSL_MSVC_STATIC_RT"] = 'MT' in str(
+                self.settings.compiler.runtime)
```
Comment (madebr) [inline recipes/openssl/3.x.x/test_package/conanfile.py:20]:
```suggestion
            cmake.definitions["OPENSSL_MSVC_STATIC_RT"] = "MT" in self.settings.compiler.runtime
```

File: recipes/openssl/3.x.x/conanfile.py [hunk 6]
```diff
@@ -0,0 +1,640 @@
+from conans import ConanFile, AutoToolsBuildEnvironment, tools
+from conans.errors import ConanInvalidConfiguration, ConanException
+import os
+import fnmatch
+import platform
```
Comment (madebr) [inline recipes/openssl/3.x.x/conanfile.py:5]: Not used either.
```suggestion
```

File: recipes/openssl/3.x.x/test_package/digest.cpp [hunk 7]
```diff
@@ -0,0 +1,89 @@
+#include <stdlib.h>
```
Comment (madebr) [inline recipes/openssl/3.x.x/test_package/digest.cpp:1]: I think this source can be renamed to `digest.c` (or `test_package.c`?).

File: recipes/openssl/config.yml (hunk 4 above)
Comment (Croydon) [reply recipes/openssl/config.yml]: Done c77da777752269d9d758a9e679adfe87838201d1

File: recipes/openssl/3.x.x/conanfile.py [hunk 8]
```diff
@@ -0,0 +1,639 @@
+from conans import ConanFile, AutoToolsBuildEnvironment, tools
+from conans.errors import ConanInvalidConfiguration, ConanException
+import os
+import fnmatch
```
Comment (madebr) [inline recipes/openssl/3.x.x/conanfile.py:4]:
```suggestion
import fnmatch
import os
import textwrap
```

File: recipes/openssl/3.x.x/conanfile.py [hunk 9]
```diff
@@ -0,0 +1,639 @@
...
+        {shared_cflag}
+        {shared_extension}
+        {cc}
+        {cxx}
+        {ar}
+        {ranlib}
+        {perlasm_scheme}
+    }},
+);
+"""
```
Comment (madebr) [inline recipes/openssl/3.x.x/conanfile.py:427]:
```suggestion
        config_template = textwrap.dedent("""\
            {targets} = (
                "{target}" => {{
                inherit_from => {ancestor},
                cflags => add("{cflags}"),
                cxxflags => add("{cxxflags}"),
                {defines}
                includes => add({includes}),
                lflags => add("{lflags}"),
                {shared_target}
                {shared_cflag}
                {shared_extension}
                {cc}
                {cxx}
                {ar}
                {ranlib}
                {perlasm_scheme}
                }},
            );
        """)
```

File: recipes/openssl/3.x.x/conanfile.py [hunk 10]
```diff
@@ -0,0 +1,639 @@
...
+            libdir = os.path.join(self.package_folder, "lib")
+            for file in os.listdir(libdir):
+                if self._is_mingw and file.endswith(".dll.a"):
+                    continue
+                if file.endswith(".a"):
+                    os.unlink(os.path.join(libdir, file))
+
+        tools.rmdir(os.path.join(self.package_folder, "lib", "pkgconfig"))
+
+    def package_info(self):
```
Comment (madebr) [inline recipes/openssl/3.x.x/conanfile.py:609]: Looks like it should also generate `openssl.pc`.

**Review by madebr — APPROVED** [review]

**Review by uilianries — DISMISSED** [review]

**Comment by uilianries** [comment]
@Croydon there was some conflicts from master, I executed `git merge -s recursive -X ours upstream/master`, which means, accept all from this PR. Also, I needed to push an empty commit, due CI, which was not able to detect solved conflicts

**Review by uilianries — APPROVED** [review]

**Comment by Croydon** [comment]
Right, because we merged #7219 quicker now

**Comment by Croydon** [comment]
@madebr Did you check for every single option that you added that they are still supported by 3.0?

**Review by madebr — APPROVED** [review]

**Comment by madebr** [comment]
Yesterday, I did a build with all all `no_*` options set to `yes`.
That's how I detected that the `openssl` util is only built when `no_stdio=False`.

**Comment by Croydon** [comment]
Thanks @madebr 

I'm fine with the state of this PR

**Review by prince-chrismc — APPROVED** [review]

**Pull request merged**